app/controllers/assessments_controller.rb

```diff
@@ -983,6 +1007,28 @@ def edit_assessment_params
       @assessment.version_penalty&.destroy
     end
 
+    if ActiveModel::Type::Boolean.new.cast(params[:unlimited_submissions]) == true
+      ass[:max_submissions] = -1
+    end
+
+    if ActiveModel::Type::Boolean.new.cast(params[:unlimited_grace_days]) == true
+      ass[:max_grace_days] = ""
+    end
+
+    if ActiveModel::Type::Boolean.new.cast(params[:use_default_late_penalty]) == true
+      ass.delete(:late_penalty_attributes)
```

Comment:
Could I get some explanation on what's going on here?

Reply:
In assessment.rb, the following relations are defined:
`belongs_to :late_penalty, class_name: "Penalty"`
`belongs_to :version_penalty, class_name: "Penalty"`
(arguably, they should be `has_one`, which I believe was one of the errors faced when trying to use the new rails defaults due to `active_record.belongs_to_required_by_default = true`)

There is also the line `accepts_nested_attributes_for :late_penalty, :version_penalty, allow_destroy: true`. This allows assessment to accept nested attributes via `late_penalty_attributes` and `version_penalty_attributes`. This creates a new Penalty record (since id is not included) and overwrites the existing Penalty if it exists.

In this case, we want to delete the attributes to avoid creating a new Penalty. We also want to destroy the existing Penalty if it exists.